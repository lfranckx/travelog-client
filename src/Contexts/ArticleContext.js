/*eslint semi: ["error", "always"]*/
import React, { Component } from 'react';

export const nullArticle = {
    author: {},
    tags: [],
};
export const nullAuthor = {};
export const nullUser = {};
export const nullUsersArticles = [];

const ArticleContext = React.createContext({
    error: null,
    article: nullArticle,
    author: nullAuthor,
    user: nullUser,
    articlesList: [],
    authorsList: [],
    usersArticles: nullUsersArticles,
    setError: () => {},
    clearError: () => {},
    setArticle: () => {},
    clearArticle: () => {},
    setAuthor: () => {},
    clearAuthor: () => {},
    setUser: () => {},
    clearUser: () => {},
    setArticlesList: () => {},
    setAuthorsList: () => {},
    setUsersArticles: () => {},
    clearUsersArticles: () => {}
});
export default ArticleContext;

export class ArticleProvider extends Component {
    state = {
        error: null,
        article: nullArticle,
        author: nullAuthor,
        user: nullUser,
        articlesList: [],
        authorsList: [],
        usersArticles: []
    };

    setError = error => {
        console.error(error);
        this.setState({ error });
    };

    clearError = () => {
        this.setState({ error: null });
    };

    setArticle = article => {
        this.setState({ article });
    };

    clearArticle = () => {
        this.setArticle(nullArticle);
    };

    setAuthor = author => {
        this.setState({ author });
    };

    clearAuthor = () => {
        this.setAuthor(nullAuthor);
    };

    setUser = user => {
        console.log('setting user:', user);
        this.setState({ user });
    };

    clearUser = () => {
        this.setState(nullUser);
    };

    setArticlesList = articlesList => {
        this.setState({ articlesList });
    };

    setAuthorsList = authorsList => {
        this.setState({ authorsList });
    };

    setUsersArticles = usersArticles => {
        this.setState({ usersArticles });
    };

    clearUsersArticles = () => {
        this.setUsersArticles(nullUsersArticles);
    };

    render() {
        const value = {
            error: this.state.error,
            article: this.state.article,
            author: this.state.author,
            user: this.state.user,
            articlesList: this.state.articlesList,
            authorsList: this.state.authorsList,
            usersArticles: this.state.usersArticles,
            setError: this.setError,
            clearError: this.clearError,
            setArticle: this.setArticle,
            clearArticle: this.clearArticle,
            setAuthor: this.setAuthor,
            clearAuthor: this.clearAuthor,
            setUser: this.setUser,
            clearUser: this.clearUser,
            setArticlesList: this.setArticlesList,
            setAuthorsList: this.setAuthorsList,
            setUsersArticles: this.setUsersArticles,
            clearUsersArticles: this.clearUsersArticles
        };
        
        return (
            <ArticleContext.Provider value={value}>
                {this.props.children}
            </ArticleContext.Provider>
        );
    }   
}