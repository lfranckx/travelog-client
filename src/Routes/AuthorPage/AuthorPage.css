.container {
    display: flex;
    flex-direction: row;
    align-items: center;
}

.author-page-container > * {
    margin: 0 5px;
}

.profile-image {
    border-radius: 50%;
    height: 60px;
    width: 60px;
}