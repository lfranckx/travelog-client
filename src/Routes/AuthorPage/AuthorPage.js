import React, { Component } from 'react';
import './AuthorPage.css';
import ArticleContext from '../../Contexts/ArticleContext'; 
import AuthorApiService from '../../Services/author-api-service';
import ArticleApiService from '../../Services/article-api-service';
import AuthorsListItem from '../../Components/AuthorsListItem/AuthorsListItem';


export default class AuthorPage extends Component {
    
    static defaultProps = {
        match: { params: {} }
    }

    static contextType = ArticleContext;

    componentDidMount() {        
        const { username } = this.props.match.params;
        this.context.clearError();
        AuthorApiService.getByUsername(username)
            .then(this.context.setAuthor)
            .catch(this.context.setError);
        ArticleApiService.getByUsername(username)
            .then(this.context.setUsersArticles)
            .catch(this.context.clearError);
    }

    componentWillUnmount() {
        this.context.clearAuthor();
        this.context.clearUsersArticles();
    }

    renderAuthor() {        
        const { author, usersArticles } = this.context;
        return (
            <>
                <section className="author-page">
                    <div className="author-page-container">
                        <img 
                            className="profile-image"
                            src={author.profile_image} 
                            alt="author-profile" />
                        <h2>{author.name}</h2>
                    </div>
                    <p>{author.about}</p>
                </section>
                <section className="authors-articles">
                    {usersArticles.map(article => 
                        <AuthorsListItem 
                            key={article.id}
                            article={article}
                        />
                    )}
                </section>
            </>
        );
    }

    render() {
        console.log('authorPage context', this.context);
        const { error, author } = this.context;        
        let content;
        if (error) {
            content = (error.error === "Author doesn't exist")
                ? <p className="error">Author not found</p>
                : <p className="error">There was an error</p>;
        } else if(!author.username) {
            content = <div className="loading">Loading...</div>;
        } else {
            content = this.renderAuthor();
        }
        return (
            <>
                {content}
            </>
        );
    }
}