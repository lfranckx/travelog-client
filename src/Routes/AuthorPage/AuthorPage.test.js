import React from 'react';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import AuthorPage from './AuthorPage';
import { ArticleProvider } from '../../Contexts/ArticleContext';

describe(`AuthorPage component`, () => {

    it(`renders form given props`, () => {
        const wrapper = mount(
            <ArticleProvider>
                <AuthorPage />
            </ArticleProvider>
        )
        expect(toJson(wrapper)).toMatchSnapshot();
    });
});