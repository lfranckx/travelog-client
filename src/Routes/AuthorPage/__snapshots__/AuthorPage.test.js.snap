// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthorPage component renders form given props 1`] = `
<ArticleProvider>
  <AuthorPage
    match={
      Object {
        "params": Object {},
      }
    }
  >
    <div
      className="loading"
    >
      Loading...
    </div>
  </AuthorPage>
</ArticleProvider>
`;
